[iOS] Mail occasionally crashes under WebKit: API::Attachment::enclosingImageData()

whsieh opened the pull request:
#### 9e3eed2f674c733b94419096a8d9859ee3278c23
<pre>
[iOS] Mail occasionally crashes under WebKit: API::Attachment::enclosingImageData()
<a href="https://bugs.webkit.org/show_bug.cgi?id=242251">https://bugs.webkit.org/show_bug.cgi?id=242251</a>
rdar://96131691

Reviewed by Chris Dumez.

The QuickLook attachment thumbnailing codepath, which uses `WKQLThumbnailLoadOperation` to request
an icon image for API-backed attachments, reads data out of the file wrapper from a background
thread. If anything simultaneously tries to access the file wrapper&apos;s content from the main thread,
the file wrapper&apos;s cached data blob will be left in a bad state (smashed, or overreleased, etc.),
potentially causing Mail to crash.

After the changes in 250772@main, we now automatically trigger QuickLook thumbnail requests when
inserting attachments via the pasteboard (i.e. drag-and-drop or copy-paste), at the same time as we
update attachment attributes (calling `enclosingImageData()` in the process on the main thread). As
a result, we hit the thread safety issue and crash described above much more frequently, resulting
in various API-attachment-related crashes that all occur when the main thread tries to read from the
file wrapper while QuickLook thumbnail operation queue is also active.

To fix this, we add a compile-time guard to ensure that access to each attachment&apos;s `m_fileWrapper`
is guarded behind a `m_fileWrapperLock`; we additionally refactor the code so that instead of
exposing a `Attachment::fileWrapper()` getter method, we have `Attachment::doWithFileWrapper()`,
which takes a C++ lambda and calls the lambda with the `NSFileWrapper`, after grabbing the lock.

* Source/WebKit/UIProcess/API/APIAttachment.cpp:
(API::Attachment::invalidate):
* Source/WebKit/UIProcess/API/APIAttachment.h:
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
(API::Attachment::setFileWrapper):
(API::Attachment::doWithFileWrapper const):

Replaces the `fileWrapper()` getter.

(API::Attachment::fileName const):
(API::Attachment::fileSizeForDisplay const):
(API::Attachment::enclosingImageData const):
(API::Attachment::enclosingImageNSData const):
(API::Attachment::isEmpty const):
(API::Attachment::createSerializedRepresentation const):
(API::Attachment::fileWrapper const): Deleted.

Adopt `doWithFileWrapper` everywhere we currently use either `m_fileWrapper` directly, or the now-
deleted `fileWrapper()` getter.

(API::Attachment::invalidateGeneratedFileWrapper): Deleted.
(API::Attachment::setFileWrapperGenerator): Deleted.

Delete support for delayed file wrapper generation; this was initially introduced in r238538 for
Apple pencil support in Mail, but was ultimately implemented using an alternate approach.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _insertAttachmentWithFileWrapper:contentType:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.mm:
(-[_WKAttachmentInfo initWithAttachment:]):

Make this take an `API::Attachment` instead of a file wrapper (and other attachment metadata) as
separate arguments. This allows us to use `doWithFileWrapper()` to access the attachment&apos;s file
wrapper behind a lock.

(-[_WKAttachmentInfo data]):
(-[_WKAttachmentInfo name]):
(-[_WKAttachmentInfo fileWrapper]):
(-[_WKAttachment info]):
(-[_WKAttachmentInfo initWithFileWrapper:filePath:mimeType:utiType:]): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::platformCloneAttachment):
(WebKit::WebPageProxy::requestThumbnail):
(WebKit::WebPageProxy::requestThumbnailWithPath):
(WebKit::WebPageProxy::requestThumbnailWithOperation): Deleted.
(WebKit::WebPageProxy::requestThumbnailWithFileWrapper): Deleted.

Drive-by refactoring: rename a couple of methods to the more succinct `requestThumbnail()`, relying
on C++ method overloading.

* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
(WebKit::WebViewImpl::writeToURLForFilePromiseProvider):
* Source/WebKit/UIProcess/QuickLookThumbnailLoader.h:
* Source/WebKit/UIProcess/QuickLookThumbnailLoader.mm:
(-[WKQLThumbnailLoadOperation initWithAttachment:identifier:]):

Make this take an `API::Attachment` instead of taking the attachment&apos;s file wrapper directly. This
allows us to use `doWithFileWrapper` to read from the file wrapper when generating the attachment
thumbnail.

(-[WKQLThumbnailLoadOperation start]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestAttachmentIcon):
(WebKit::WebPageProxy::updateAttachmentAttributes):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromFilePath):
(WebKit::WebPageProxy::registerAttachmentsFromSerializedData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _prepareToDragPromisedAttachment:]):
(createItemProvider):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:
(TestWebKitAPI::TEST):

Drive-by fix: make `WKAttachmentTestsMac.DraggingAttachmentBackedImagePreservesRangedSelection` pass
when running API tests locally, when the TestWebKitAPI window is not ordered to the front. When
running this test from command line from Terminal, the API test&apos;s window ends up behind the terminal
window, which causes the drag session to never end.

Canonical link: <a href="https://commits.webkit.org/252118@main">https://commits.webkit.org/252118@main</a>
</pre>

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
```diff
@@ -124,61 +124,69 @@ static BOOL isDeclaredOrDynamicTypeIdentifier(NSString *type)
-    if (!data)
-        return nullptr;
+        data = [m_fileWrapper regularFileContents];
+        if (!data)
```
Nit: this check could be outside the lock scope.

whsieh replied:
Fixed!

cdumez commented on Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
```diff
@@ -124,61 +124,69 @@ static BOOL isDeclaredOrDynamicTypeIdentifier(NSString *type)
-    if (!serializedData)
-        return nullptr;
+        serializedData = [NSKeyedArchiver archivedDataWithRootObject:m_fileWrapper.get() requiringSecureCoding:YES error:nullptr];
+        if (!serializedData)
```
nit: this check could be outside the lock scope.

whsieh replied:
Good catch! Fixed.

cdumez commented on Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
```diff
@@ -10321,11 +10321,13 @@ - (void)_didExitFullscreen
-            completionHandler(nil, [NSError errorWithDomain:WKErrorDomain code:WKErrorUnknown userInfo:nil]);
+    [item registerDataRepresentationForTypeIdentifier:utiType visibility:NSItemProviderRepresentationVisibilityAll loadHandler:[attachment](void (^completionHandler)(NSData *, NSError *)) -> NSProgress * {
+        attachment->doWithFileWrapper([&](NSFileWrapper *fileWrapper) {
+            if (auto nsData = RetainPtr { fileWrapper.serializedRepresentation })
```
nit: Would this work?
`if (RetainPtr nsData = fileWrapper.serializedRepresentation)`

whsieh replied:
Ah, so I originally tried exactly that, but got this error:

```
/Volumes/Data/worker/iOS-15-Simulator-Build-EWS/build/Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:10326:27: error: no viable constructor or deduction guide for deduction of template arguments of 'RetainPtr'
```

cdumez reviewed (approved):
LGTM

whsieh committed:
[iOS] Mail occasionally crashes under WebKit: API::Attachment::enclosingImageData()
https://bugs.webkit.org/show_bug.cgi?id=242251
rdar://96131691

Reviewed by Chris Dumez.

The QuickLook attachment thumbnailing codepath, which uses `WKQLThumbnailLoadOperation` to request
an icon image for API-backed attachments, reads data out of the file wrapper from a background
thread. If anything simultaneously tries to access the file wrapper's content from the main thread,
the file wrapper's cached data blob will be left in a bad state (smashed, or overreleased, etc.),
potentially causing Mail to crash.

After the changes in 250772@main, we now automatically trigger QuickLook thumbnail requests when
inserting attachments via the pasteboard (i.e. drag-and-drop or copy-paste), at the same time as we
update attachment attributes (calling `enclosingImageData()` in the process on the main thread). As
a result, we hit the thread safety issue and crash described above much more frequently, resulting
in various API-attachment-related crashes that all occur when the main thread tries to read from the
file wrapper while QuickLook thumbnail operation queue is also active.

To fix this, we add a compile-time guard to ensure that access to each attachment's `m_fileWrapper`
is guarded behind a `m_fileWrapperLock`; we additionally refactor the code so that instead of
exposing a `Attachment::fileWrapper()` getter method, we have `Attachment::doWithFileWrapper()`,
which takes a C++ lambda and calls the lambda with the `NSFileWrapper`, after grabbing the lock.

* Source/WebKit/UIProcess/API/APIAttachment.cpp:
(API::Attachment::invalidate):
* Source/WebKit/UIProcess/API/APIAttachment.h:
* Source/WebKit/UIProcess/API/Cocoa/APIAttachmentCocoa.mm:
(API::Attachment::setFileWrapper):
(API::Attachment::doWithFileWrapper const):

Replaces the `fileWrapper()` getter.

(API::Attachment::fileName const):
(API::Attachment::fileSizeForDisplay const):
(API::Attachment::enclosingImageData const):
(API::Attachment::enclosingImageNSData const):
(API::Attachment::isEmpty const):
(API::Attachment::createSerializedRepresentation const):
(API::Attachment::fileWrapper const): Deleted.

Adopt `doWithFileWrapper` everywhere we currently use either `m_fileWrapper` directly, or the now-
deleted `fileWrapper()` getter.

(API::Attachment::invalidateGeneratedFileWrapper): Deleted.
(API::Attachment::setFileWrapperGenerator): Deleted.

Delete support for delayed file wrapper generation; this was initially introduced in r238538 for
Apple pencil support in Mail, but was ultimately implemented using an alternate approach.

* Source/WebKit/UIProcess/API/Cocoa/WKWebView.mm:
(-[WKWebView _insertAttachmentWithFileWrapper:contentType:completion:]):
* Source/WebKit/UIProcess/API/Cocoa/_WKAttachment.mm:
(-[_WKAttachmentInfo initWithAttachment:]):

Make this take an `API::Attachment` instead of a file wrapper (and other attachment metadata) as
separate arguments. This allows us to use `doWithFileWrapper()` to access the attachment's file
wrapper behind a lock.

(-[_WKAttachmentInfo data]):
(-[_WKAttachmentInfo name]):
(-[_WKAttachmentInfo fileWrapper]):
(-[_WKAttachment info]):
(-[_WKAttachmentInfo initWithFileWrapper:filePath:mimeType:utiType:]): Deleted.
* Source/WebKit/UIProcess/Cocoa/WebPageProxyCocoa.mm:
(WebKit::WebPageProxy::platformCloneAttachment):
(WebKit::WebPageProxy::requestThumbnail):
(WebKit::WebPageProxy::requestThumbnailWithPath):
(WebKit::WebPageProxy::requestThumbnailWithOperation): Deleted.
(WebKit::WebPageProxy::requestThumbnailWithFileWrapper): Deleted.

Drive-by refactoring: rename a couple of methods to the more succinct `requestThumbnail()`, relying
on C++ method overloading.

* Source/WebKit/UIProcess/Cocoa/WebViewImpl.mm:
(WebKit::WebViewImpl::writeToURLForFilePromiseProvider):
* Source/WebKit/UIProcess/QuickLookThumbnailLoader.h:
* Source/WebKit/UIProcess/QuickLookThumbnailLoader.mm:
(-[WKQLThumbnailLoadOperation initWithAttachment:identifier:]):

Make this take an `API::Attachment` instead of taking the attachment's file wrapper directly. This
allows us to use `doWithFileWrapper` to read from the file wrapper when generating the attachment
thumbnail.

(-[WKQLThumbnailLoadOperation start]):
* Source/WebKit/UIProcess/WebPageProxy.cpp:
(WebKit::WebPageProxy::requestAttachmentIcon):
(WebKit::WebPageProxy::updateAttachmentAttributes):
(WebKit::WebPageProxy::registerAttachmentIdentifierFromFilePath):
(WebKit::WebPageProxy::registerAttachmentsFromSerializedData):
* Source/WebKit/UIProcess/WebPageProxy.h:
* Source/WebKit/UIProcess/ios/WKContentViewInteraction.mm:
(-[WKContentView _prepareToDragPromisedAttachment:]):
(createItemProvider):
* Tools/TestWebKitAPI/Tests/WebKitCocoa/WKAttachmentTests.mm:
(TestWebKitAPI::TEST):

Drive-by fix: make `WKAttachmentTestsMac.DraggingAttachmentBackedImagePreservesRangedSelection` pass
when running API tests locally, when the TestWebKitAPI window is not ordered to the front. When
running this test from command line from Terminal, the API test's window ends up behind the terminal
window, which causes the drag session to never end.

Canonical link: https://commits.webkit.org/252118@main
